Add support in plugin config for accessing host ipc and pid namespace.

anusha-ragunathan opened the pull request:
Plugins might need access to host ipc and pid namespaces. A good usecase is
a volume plugin running iscsi multipath commands that need access to
host kernel locks.
Tested with a custom plugin (aragunathan/global-net-plugin-full) that's
built with `"ipchost" : true` and `"pidhost": true` in config.json. Observed using
`readlink /proc/self/ns/ipc` and `readlink /proc/self/ns/pid`that plugin and host 
have the same ns.

Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>

anusha-ragunathan committed:
Add support in plugin config for accessing host ipc namespace.

Plugins might need access to host ipc namespace. A good usecase is
a volume plugin running iscsi multipath commands that need access to
host kernel locks.
Tested with a custom plugin (aragunathan/global-net-plugin-full) that's
built with `"ipchost" : true` in config.json. Observed using
`readlink /proc/self/ns/ipc` that plugin and host have the same ns.

Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>

cpuguy83 commented:
```
00:47:13 The result of hack/generate-swagger-api.sh differs
00:47:13 
00:47:13  M api/types/plugin.go
00:47:13 
00:47:13 Please update api/swagger.yaml with any api changes, then 
00:47:13 run `hack/generate-swagger-api.sh`.
```

cpuguy83 commented:
Do we need some kind of version bump for these kinds of changes?
A plugin that requires this won't work on 17.0.x or 1.13.x

anusha-ragunathan commented:
@cpuguy83 : Yes, a plugin config versioning PR is coming soon.

anusha-ragunathan committed:
Add pid host support

Tested using global-net-plugin-ipc which sets PidHost in config.json.

Plugins might need access to host pid namespace. Add support for that.
Tested using aragunathan/global-net-plugin-ipc which sets "pidhost" in
config.json. Observed using `readlink /proc/self/ns/pid` that plugin and
host have the same ns.

Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>

vieux commented:
LGTM

tiborvass commented:
I would prefer merging the version PR at the same time as this PR, since we're close to code freeze for 17.04.

anusha-ragunathan commented:
@tiborvass : Added version commit to this PR. PTAL

anusha-ragunathan commented:
Currently, the commit performs version check during `plugin enable`. It's better to have the check during `plugin install` when computing privileges; the only downside is that API users can skip `get privileges` during an install.

thaJeztah commented:
@anusha-ragunathan @vieux the 17.04 branch was created; is this still intended for 17.04 or should I move it to 17.05?

anusha-ragunathan commented:
As far as criticality of this PR, it is not very critical for 17.04. However, if we are still open to taking fixes in, then we should consider it for 17.04. Else, I'm okay moving to 17.05.

anusha-ragunathan commented:
@cpuguy83 @justincormack @tonistiigi  : WDYT about versioning the plugin config?

anusha-ragunathan commented:
Here's why versioning plugin config is important:
Lets say 
- in 17.03, a plugin was released as johndoe/foo:v1
- in 17.05, the plugin uses the new ipchost config setting and author publishes johndoe/foo:v2

If we embed the `DockerVersion` as part of the plugin config when it is created, users can do a 
`docker plugin inspect` and know which version the plugin was built on. This can be helpful in cases where users are running a new plugin config on older docker releases and confused at the resulting behavior.

In this case, johndoe/foo:v2 was built on 17.05 using ipchost and is running on 17.03
- Plugin config's DockerVersion: 17.05
- Docker version: 17.03

There's no guarantee that if the plugin config `DockerVersion` > Docker Version, the settings will work.
Ofcourse, if the johndoe/foo:v2 has no config settings newly added in 17.05, things will work fine.

anusha-ragunathan committed:
Embed DockerVersion in plugin config.

Embedding DockerVersion in plugin config when the plugin is created,
enables users to do a docker plugin inspect and know which version
the plugin was built on. This is helpful in cases where users are
running a new plugin on older docker releases and confused at
unexpected behavior.

By embedding DockerVersion in the config, we claim that there's no
guarantee that if the plugin config's DockerVersion is greater that
the version of the docker engine the plugin is executed against, the
plugin will work as expected.

For example, lets say:
- in 17.03, a plugin was released as johndoe/foo:v1
- in 17.05, the plugin uses the new ipchost config setting and author
publishes johndoe/foo:v2

In this case, johndoe/foo:v2 was built on 17.05 using ipchost, but is
running on docker-engine version 17.03. Since 17.05 > 17.03, there's
no guarantee that the plugin will work as expected. Ofcourse, if the
plugin did not use newly added config settings (ipchost in this case)
in 17.05, it would work fine in 17.03.

Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>

tonistiigi commented on api/types/plugin.go:
```diff
@@ -42,6 +42,10 @@ type PluginConfig struct {
 
+	// Docker Version used to create the plugin
+	// Required: true
+	DockerVersion string `json:"DockerVersion"`
```
Should the new fields be marked `omitempty` so they don't show up as empty values for old plugins?

anusha-ragunathan replied:
I dont feel too strongly either way. This case is applicable only for plugins built using 17.03/17.04 and running on versions >= 17.05.

anusha-ragunathan replied:
I'm thinking of a case where users wants an inventory of all plugins versions. If we have an empty string, they could use `docker plugin inspect <id> --format {{.Config.DockerVersion}}` and if there are ones returning empty know that they were built on 17.03/17.04. If we omit it from the json, then `plugin inspect` will error out.  This makes me lean more towards having an empty version, rather than omitting it. WDYT?

tonistiigi replied:
I don't think that format case errors as it still uses the golang definition.

anusha-ragunathan replied:
@tonistiigi : Updated to add `omitempty`

tonistiigi commented:
LGTM

tiborvass commented:
I want to review this, please don't merge it without my review.

cpuguy83 commented:
I'm not convinced that using `DockerVersion` as the plugin API version is the right approach here. It seems ok now because we'll probably be iterating on this for the next few versions, but once it's stabilized we just keep incrementing?

anusha-ragunathan commented:
Embedding `DockerVersion` in the plugin config is metadata. It is not versioning the plugin config. It's similar to what we do for images and is the source of truth for where the plugin was created.

cpuguy83 commented:
Apologies I misunderstood the purpose!

tiborvass commented:
So we've been whiteboarding all kinds of solutions and variants within them.

Here are some dimensions of solutions:
1. Versioning: (a) none, (b) plugin-specific, (c) docker API version, (d) docker version
2. Semver: (y) yes, (n) no
3. Per-field versioning: (y) yes, (n) no
4. Auto-detect unknown fields: (y) yes, (n) no

TLDR We agreed on `1d 3n 4y`. So LGTM as-is and we can have a follow up for 4.

If we don't have versioning, users won't know how to debug a new plugin running on older docker that's not behaving as expected.
Auto-detecting unknown fields is a good step, but not sufficient, because it requires users to look up when did that new field appear.
So we could automatically version each field by annotating it with a version in which they were added. But that would be a new thing in docker and was deemed to complex given the other simpler solutions.
So now that we've established we want a version, should it be semver? Semver doesn't buy us anything, because if we're running a 1.1.0 plugin on a 1.0.0. runtime, we don't want to error out if the 1.1.0-specific fields are not part of the plugin, so we definitely need the auto-detection of unknown fields. Having a plugin-specific version brings confusion, because plugin interfaces also have a version (authz, volumes, network). We're left with choosing between docker API version and docker version: since we're talking about users, they only should know about docker version.

Also keep in mind this is informational only and nothing is being enforced. In another PR, we can enable auto-detection of unknown fields, and if some are detected then warn about the version mismatch.
